Issue #17 and Issue #13

judgej opened the pull request:
Being more consistent with the amount: getAmount() will always return a formatted string. Some additional checks on the format of the value passed in (floats, integers and strings are allowed, so long as their value is unambiguous - integers still not accepted for currencies that support decimals, for legacy reasons).

Zero amounts now explicitly allowed, and negative amounts checked for. However, both of these can be reversed by overriding the appropriate properties.

Decimal point is still always "." regardless of locale. This is derived from the PHP format for floats, which always uses a ".". Any gateway that needs to send other decimal characters to the remote gateway, will need to overwrite getAmount().

kayladnls commented on src/Omnipay/Common/Message/AbstractRequest.php:
```diff
@@ -271,21 +281,67 @@ public function setCardReference($value)
+        }
+
+        return (float)$value;
+    }
```
I'm thinking this function would be better suited for Helper.php.

judgej replied:
I'll take a look at doing this. I could imagine it would be easier alter if amounts can be supplied in other formats, such as a currency-value or cash object.

kayladnls replied:
That's a very good point. Might be worth adding in https://github.com/mathiasverraes/money in V3

judgej replied:
The `InvalidRequestException` seems like the wrong exception to raise on a validation error in the Helper class. `ValidationException` seems more appropriate. Possibly `\InvalidArgumentException` in the first instance (if the supplied type is not a int/string/float).

Can this Helper method call those exceptions directly, or do ALL exceptions raised explicitly by OmniPay logic need to be extended into the OmniPay namespace?

kayladnls commented on tests/Omnipay/Common/Message/AbstractRequestTest.php:
```diff
@@ -94,11 +106,15 @@ public function testGetAmountNoDecimals()
         $this->assertSame($this->request, $this->request->setAmount('136.5'));
         $this->assertSame($this->request, $this->request->setCurrency('JPY'));
-        $this->assertSame('137', $this->request->getAmount());
+        $this->request->getAmount();
```
Shouldn't this be an assertion?

kayladnls commented:
Other than that one comment, this looks wonderful. Thank you for your work on this, as always @judgej

judgej commented:
I think this may need a few changes still, but see what you think. This is what I have done:

1. Moved toFloat() to the Helper class.
2. Changed the Helper version exceptions from `\Omnipay\Common\Exception\InvalidRequestException` to `\InvalidArgumentException`
3. The original AbstractMessage::toFloat() now calls the helper version but re-throws `InvalidArgumentException` as `InvalidRequestException` to support legacy code.
4. Test coverage for the new `toFloat()`.

What may need to change:

1. Create new exception `\Omnipay\Common\Exception\InvalidArgumentException`
2. Remove the rethrowing of the exception in the old `toFloat()`.
3. Deprecate the old `toFloat()` and call the new Helper version. It is likely that very few, if any, drivers call it directly anyway. Deprecation would give some time for those that do to be changed, without breaking production code.
